clusterloader2/pkg/measurement/common/ooms_tracker.go

```diff
@@ -0,0 +1,216 @@
+)
+
+func init() {
+	if err := measurement.Register(clusterOomsTrackerName, createClusterOomsTrackerMeasurement); err != nil {
+		klog.Fatalf("Cannot register %s: %v", clusterOomsTrackerName, err)
+	}
+}
+
+func createClusterOomsTrackerMeasurement() measurement.Measurement {
+	return &clusterOomsTrackerMeasurement{}
+}
+
+type clusterOomsTrackerMeasurement struct {
+	selector  *measurementutil.ObjectSelector
+	msgRegex  *regexp.Regexp
+	isRunning bool
+	stopCh    chan struct{}
+	ooms      []oomEvent
+	oomsLock  sync.Mutex
+}
+
+type oomEvent struct {
+	Node          string    `json:"node"`
+	Process       string    `json:"process"`
+	ProcessMemory string    `json:"memory"`
+	ProcessID     int       `json:"pid"`
+	Time          time.Time `json:"time"`
+}
+
+func (m *clusterOomsTrackerMeasurement) Execute(config *measurement.Config) ([]measurement.Summary, error) {
+	clusterOomsTrackerEnabled, err := util.GetBoolOrDefault(config.Params, clusterOomsTrackerEnabledParamName, false)
+	if err != nil {
+		return nil, goerrors.Errorf("problem with getting %s param: %v", clusterOomsTrackerEnabledParamName, err)
+	}
+	if !clusterOomsTrackerEnabled {
+		klog.Info("skipping tracking of OOMs in the cluster")
+		return nil, nil
+	}
+
+	action, err := util.GetString(config.Params, "action")
+	if err != nil {
+		return nil, goerrors.Errorf("problem with getting %s param: %v", "action", err)
+	}
+
+	switch action {
+	case "start":
+		if err = m.start(config.ClusterFramework.GetClientSets().GetClient()); err != nil {
+			return nil, goerrors.Errorf("starting cluster OOMs measurement problem: %v", err)
+		}
+		return nil, nil
+	case "gather":
+		m.oomsLock.Lock()
+		defer m.oomsLock.Unlock()
+		summary, err := m.gather()
+		if err != nil {
+			return nil, goerrors.Errorf("gathering cluster OOMs measurement problem: %v", err)
+		}
+		return summary, m.validate()
+	default:
+		return nil, goerrors.Errorf("unknown action %v", action)
+	}
+}
+
+func (m *clusterOomsTrackerMeasurement) Dispose() {
+	m.stop()
+}
+
+func (m *clusterOomsTrackerMeasurement) String() string {
+	return clusterOomsTrackerName
+}
+
+func (m *clusterOomsTrackerMeasurement) start(c clientset.Interface) error {
+	if m.isRunning {
+		klog.Infof("%s: cluster OOMs tracking measurement already running", m)
+		return nil
+	}
+	klog.Infof("%s: starting cluster OOMs tracking measurement...", m)
+	m.initFields()
+	// Watching for OOM events from node-problem-detector below.
+	i := informer.NewInformer(
+		c,
+		"events",
+		m.selector,
+		m.handleOomEvent,
+	)
+	if err := informer.StartAndSync(i, m.stopCh, informerTimeout); err != nil {
+		return goerrors.Errorf("problem with OOM events informer starting: %v", err)
+	}
+	return nil
+}
+
+func (m *clusterOomsTrackerMeasurement) initFields() {
+	m.isRunning = true
+	m.stopCh = make(chan struct{})
+	m.ooms = make([]oomEvent, 0)
```

Comment:
NIT: is there actually a benefit to create slice at all?

Reply:
It's not necessary as one can append to a `nil` slice in Golang as well it seems. The only benefit I see is that when there are no OOMs, `PrettyPrintJSON` will print out a JSON summary like that:
```
{
    ooms: []
}
```
instead of
```
{
    ooms: null
}
```
but it's super minor stuff.

Reply:
After offline discussion we agreed to leave it as it is. Done.

Reply:
Yep, pretty printing JSON is a valid reason.